cosmwasm/contracts/wormchain-accounting/src/contract.rs

```diff
@@ -0,0 +1,469 @@
+            )
+        })
+        .filter_map(Result::transpose)
+        .collect::<anyhow::Result<Vec<_>>>()
+        .context("failed to handle `Observation`")?;
+
+    Ok(Response::new()
+        .add_attribute("action", "submit_observations")
+        .add_attribute("owner", info.sender)
+        .add_events(events))
+}
+
+fn handle_observation(
+    mut deps: DepsMut<WormholeQuery>,
+    o: Observation,
+    guardian_set_index: u32,
+    quorum: usize,
+    sig: Signature,
+) -> anyhow::Result<Option<Event>> {
+    if accounting::has_transfer(deps.as_ref(), o.key.clone()) {
+        bail!("transfer for key \"{}\" already committed", o.key);
+    }
+
+    let key = PENDING_TRANSFERS.key(o.key.clone());
+    let mut pending = key
+        .may_load(deps.storage)
+        .map(Option::unwrap_or_default)
+        .context("failed to load `PendingTransfer`")?;
+    let data = match pending
+        .iter_mut()
+        .find(|d| d.guardian_set_index() == guardian_set_index && d.observation() == &o)
+    {
+        Some(d) => d,
+        None => {
+            pending.push(Data::new(o.clone(), guardian_set_index));
+            let back = pending.len() - 1;
+            &mut pending[back]
+        }
+    };
+
+    data.add_signature(sig)?;
+
+    if data.signatures().len() < quorum {
+        // Still need more signatures so just save the pending transfer data and exit.
+        key.save(deps.storage, &pending)
+            .context("failed to save pending transfers")?;
+
+        return Ok(None);
+    }
+
+    let (msg, _) = serde_wormhole::from_slice_with_payload(&o.payload)
+        .context("failed to parse observation payload")?;
+    let tx_data = match msg {
+        Message::Transfer {
+            amount,
+            token_address,
+            token_chain,
+            recipient_chain,
+            ..
+        }
+        | Message::TransferWithPayload {
+            amount,
+            token_address,
+            token_chain,
+            recipient_chain,
+            ..
+        } => transfer::Data {
+            amount: Uint256::from_be_bytes(amount.0),
+            token_address: TokenAddress::new(token_address.0),
+            token_chain: token_chain.into(),
+            recipient_chain: recipient_chain.into(),
+        },
+        _ => bail!("Unknown tokenbridge payload"),
+    };
+
+    let emitter_chain = o.key.emitter_chain();
+
+    let tokenbridge_addr = TOKENBRIDGE_ADDR
+        .load(deps.storage)
+        .context("failed to load tokenbridge addr")?;
+
+    let registered_emitter: Vec<u8> = deps
+        .querier
+        .query_wasm_smart(
+            tokenbridge_addr,
+            &tokenbridge::msg::QueryMsg::ChainRegistration {
+                chain: emitter_chain,
+            },
+        )
+        .context("failed to query chain registration")?;
+    ensure!(
+        *registered_emitter == **o.key.emitter_address(),
+        "unknown emitter address"
+    );
+
+    accounting::commit_transfer(
```

Comment:
Should the event from this method be propagated to the `Event` down below with `add_event` like in `instantiate`? Some of the same info from the `commit_transfer` `Event` is shared with this method's event.

Reply:
I'm inclined to say no because the information is repeated, as you mentioned.  Someone listening for this event can get all the details by parsing out the `payload` field.

That said, I'm also ok with adding both events.  I don't feel that strongly either way.

Reply:
I'm cool the way it is. Someone else integrating w/ the accounting package can choose to use this event or their own custom event. In this case, the wormchain accounting smart contract chooses to use its own Event.